app/code/Magento/InventoryCatalog/Test/Integration/InventoryCatalog/Model/Source/Validator/PreventDisableValidatorTest.php

```diff
@@ -0,0 +1,117 @@
+
+    /**
+     * @param string $field
+     * @param $value
+     * @param array $expectedErrorData
+     * @dataProvider failedValidationDataProvider
+     * @throws \Exception
+     */
+    public function testFailedValidationOnUpdate(string $field, $value, array $expectedErrorData)
+    {
+        $data = $this->validData;
+        $data[$field] = $value;
+
+        $sourceCode = 'default';
+        $serviceInfo = [
+            'rest' => [
+                'resourcePath' => self::RESOURCE_PATH . '/' . $sourceCode,
+                'httpMethod' => Request::HTTP_METHOD_PUT,
+            ],
+            'soap' => [
+                'service' => self::SERVICE_NAME,
+                'operation' => self::SERVICE_NAME . 'Save',
+            ],
+        ];
+        $this->webApiCall($serviceInfo, $data, $expectedErrorData);
+    }
+
+    /**
+     * SuppressWarnings was added due to a tests on different fail types and big size of data provider
+     *
+     * @return array
+     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
+     */
+    public function failedValidationDataProvider(): array
+    {
+        return [
+            'disabled_' . SourceInterface::SOURCE_CODE => [
+                SourceInterface::SOURCE_CODE,
+                false,
+                [
+                    'message' => 'Validation Failed',
+                    'errors' => [
+                        [
+                            'message' => 'Default source can not be disabled.',
+                            'parameters' => [],
+                        ],
+                    ],
+                ],
+            ],
+        ];
+    }
+
+    /**
+     * @param array $serviceInfo
+     * @param array $data
+     * @param array $expectedErrorData
+     * @return void
+     * @throws \Exception
+     */
+    private function webApiCall(array $serviceInfo, array $data, array $expectedErrorData)
+    {
+        try {
+            $this->_webApiCall($serviceInfo, ['source' => $data]);
+            $this->fail('Expected throwing exception');
+        } catch (\Exception $e) {
+            if (TESTS_WEB_API_ADAPTER === self::ADAPTER_REST) {
+                self::assertEquals($expectedErrorData, $this->processRestExceptionResult($e));
+                self::assertEquals(Exception::HTTP_BAD_REQUEST, $e->getCode());
+            } elseif (TESTS_WEB_API_ADAPTER === self::ADAPTER_SOAP) {
+                $this->assertInstanceOf('SoapFault', $e);
+                $expectedWrappedErrors = [];
+                foreach ($expectedErrorData['errors'] as $error) {
+                    // @see \Magento\TestFramework\TestCase\WebapiAbstract::getActualWrappedErrors()
+                    $expectedWrappedErrors[] = [
+                        'message' => $error['message'],
+                        'params' => $error['parameters'],
+                    ];
+                }
+                $this->checkSoapFault($e, $expectedErrorData['message'], 'env:Sender', [], $expectedWrappedErrors);
```

Comment:
Pls check run this testcase in SOAP mode on local machine

Reply:
sure, I'll try.